presto-common/src/main/java/com/facebook/presto/common/PrestoVersion.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.common;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static java.util.Objects.requireNonNull;
+
+public class PrestoVersion
+        implements Comparable<PrestoVersion>
+{
+    private static final Pattern VERSION_REGEX = Pattern.compile("^(0\\.([1-9][0-9]*))(\\.([1-9][0-9]*))?(\\.*)?(-([2][0-9]{7}\\.[0-2][0-9]{5}-[0-9]{1,}|SNAPSHOT).*)?$");
+    private int majorVersion;
+    private int minorVersion;
+    private String snapshotVersion;
+    private String version;
+
+    public PrestoVersion(String version)
+    {
+        Matcher matcher = VERSION_REGEX.matcher(version);
+        this.majorVersion = Integer.parseInt(Optional.ofNullable(matcher.lookingAt() ? matcher.group(2) : null).orElse("0"));
+        this.minorVersion = Integer.parseInt(Optional.ofNullable(matcher.lookingAt() ? matcher.group(4) : null).orElse("0"));
+        this.snapshotVersion = Optional.ofNullable(matcher.lookingAt() ? matcher.group(7) : null).orElse("");
```

Comment:
Is the call to repeated `lookingAt` necessary?. can't you check for `matcher.group != null`

Reply:
If we use `matcher.group(2)` and `matcher.group(2)` does not exist, the code throws `IllegalStateException`

Reply:
can you double check the behavior. from the javadocs.

> If the match was successful but the group specified failed to match any part of the input sequence, then null is returned
> Throws:
[IllegalStateException](https://docs.oracle.com/javase/7/docs/api/java/lang/IllegalStateException.html) - If no match has yet been attempted, or if the previous match operation failed

Reply:
I've updated the code to use `matcher.find()` instead of `lookingAt()` since it seemed like a better solution.